File: src/profileUrl.ts
import { InvalidProfileUrlError } from "./errors";

// Accepts e.g. https://www.linkedin.com/in/john-doe-12345/ or linkedin.com/in/john-doe (with or without scheme/trailing slash/query string).
export function extractPublicIdentifier(profileUrl: string): string {
  let url: URL;
  try {
    url = new URL(profileUrl.startsWith("http") ? profileUrl : `https://${profileUrl}`);
  } catch {
    throw new InvalidProfileUrlError(profileUrl);
  }

  if (!url.hostname.endsWith("linkedin.com")) {
    throw new InvalidProfileUrlError(profileUrl);
  }

  const match = url.pathname.match(/\/in\/([^/]+)/);
  if (!match) {
    throw new InvalidProfileUrlError(profileUrl);
  }

  return decodeURIComponent(match[1]);
}

File: app/profile_url.py
import re
from urllib.parse import unquote, urlparse

from .errors import InvalidProfileUrlError

_IN_PATH = re.compile(r"/in/([^/]+)")


def extract_public_identifier(profile_url: str) -> str:
    """Accepts https://www.linkedin.com/in/john-doe/, linkedin.com/in/john-doe?x=1, etc.

    The linkedin.com host check is deliberate: this value comes straight from a query parameter,
    so without it the service would happily fetch any URL a caller supplies (SSRF).
    """
    candidate = profile_url if profile_url.startswith("http") else f"https://{profile_url}"
    parsed = urlparse(candidate)

    hostname = (parsed.hostname or "").lower()
    if not (hostname == "linkedin.com" or hostname.endswith(".linkedin.com")):
        raise InvalidProfileUrlError(profile_url)

    match = _IN_PATH.search(parsed.path)
    if not match:
        raise InvalidProfileUrlError(profile_url)

    return unquote(match.group(1))
